sw/device/tests/dif/dif_aes_smoketest.c

```diff
@@ -0,0 +1,152 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_aes.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+// The following plaintext, key and ciphertext are extracted from Appendix C of
+// the Advanced Encryption Standard (AES) FIPS Publication 197 available at
+// https://www.nist.gov/publications/advanced-encryption-standard-aes
+
+#define KEY_SIZE_IN_BYTES 32
+#define TEXT_SIZE_IN_BYTES 16
+#define TEXT_SIZE_IN_WORDS (TEXT_SIZE_IN_BYTES / 4)
+
+static const uint8_t kPlainText[TEXT_SIZE_IN_BYTES] = {
+    0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77,
+    0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff,
+};
+
+static const uint8_t kKey[KEY_SIZE_IN_BYTES] = {
+    0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a,
+    0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15,
+    0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
+};
+
+static const uint8_t kCipherTextGold[TEXT_SIZE_IN_BYTES] = {
+    0x8e, 0xa2, 0xb7, 0xca, 0x51, 0x67, 0x45, 0xbf,
+    0xea, 0xfc, 0x49, 0x90, 0x4b, 0x49, 0x60, 0x89,
+};
+
+// The mask share, used to mask kKey. Note that the masking should not be done
+// manually. Software is expected to get the key in two shares right from the
+// beginning.
+static const uint8_t kKeyShare1[KEY_SIZE_IN_BYTES] = {
+    0x0f, 0x1f, 0x2f, 0x3F, 0x4f, 0x5f, 0x6f, 0x7f, 0x8f, 0x9f, 0xaf,
+    0xbf, 0xcf, 0xdf, 0xef, 0xff, 0x0a, 0x1a, 0x2a, 0x3a, 0x4a, 0x5a,
+    0x6a, 0x7a, 0x8a, 0x9a, 0xaa, 0xba, 0xca, 0xda, 0xea, 0xfa,
+};
+
+const test_config_t kTestConfig;
+
+static bool aes_input_ready(const dif_aes_t *aes) {
+  bool status;
+  CHECK(dif_aes_get_status(aes, kDifAesStatusInputReady, &status) == kDifAesOk);
+
+  return status;
+}
+
+static bool aes_output_valid(const dif_aes_t *aes) {
+  bool status;
+  CHECK(dif_aes_get_status(aes, kDifAesStatusOutputValid, &status) ==
+        kDifAesOk);
+
+  return status;
+}
+
+bool test_main(void) {
+  dif_aes_t aes;
+
+  LOG_INFO("Running AES test");
+
+  // Initialise AES.
+  dif_aes_params_t params = {
+      .base_addr = mmio_region_from_addr(TOP_EARLGREY_AES_BASE_ADDR),
+  };
+  CHECK(dif_aes_init(params, &aes) == kDifAesOk);
+  CHECK(dif_aes_reset(&aes) == kDifAesResetOk);
+
+  // Mask the key. Note that this should not be done manually. Software is
+  // expected to get the key in two shares right from the beginning.
+  uint8_t key_share0[KEY_SIZE_IN_BYTES];
+  for (int i = 0; i < KEY_SIZE_IN_BYTES; ++i) {
+    key_share0[i] = kKey[i] ^ kKeyShare1[i];
+  }
+
+  // "Convert" key share byte arrays to `dif_aes_key_share_t`.
+  dif_aes_key_share_t key;
+  memcpy(&key.share0[0], &key_share0[0], KEY_SIZE_IN_BYTES);
```

Comment:
For array types, `&foo[0]` is redundant, since the array will just decay into a pointer to itself.

Reply:
Yes, but I think it is an acceptable stylistic approach (I actually prefer it as it implies dealing with an array).

Reply:
I am also with @silvestrst here.